### ISSUE TITLE
Rancher 2.5.9 uses flannel:v0.13.0 which causes iptables segfault in RHEL 8.4

### ISSUE THREAD

**Issue by rajha-korithrien** [rancher/rancher#34604]
**Rancher Server Setup**
- Rancher version: 2.5.9
- Installation option: Helm Chart, RKE1, but the cluster we observe the issue in was entirely created in the Rancher UI

**Information about the Cluster**
- Kubernetes version: 1.20.9
- Cluster Type (Local/Downstream): Downstream, Custom

**Describe the bug**
Nodes running RHEL 8.4 4.18.0-305.12.1.el8_4.x86_64 have the following in dmesg
```
[Wed Sep  1 16:52:49 2021] iptables-save[999039]: segfault at 80 ip 00007ff2b2f4d964 sp 00007ffff48cca18 error 4 in libnftnl.so.11.2.0[7ff2b2f48000+19000]
[Wed Sep  1 16:52:49 2021] Code: 83 c4 08 5b 5d 41 5c 41 5d c3 0f 1f 40 00 48 83 c4 08 31 c0 5b 5d 41 5c 41 5d c3 66 66 2e 0f 1f 84 00 00 00 00 00 f3 0f 1e fa <48> 8b 87 80 00 00 00 48 83 ef 80 48 39 f8 74 1b 85 f6 75 0c eb 18
```
and can not pass overlay traffic.

**To Reproduce**
1) Use RKE1 to create a local cluster, use Helm to install Rancher 2.5.9 into local cluster
2) Use the Rancher UI to create a downstream cluster with Canal CNI and Project Isolation enabled, add RHEL 8.4 nodes with a minimal install and Docker CE 20.10.8
3) Deploy workloads to the RHEL 8.4 nodes that pass traffic via the overlay
4) Observe that nodes can not communicate via overlay network
5) Observe that you can not use the Rancher UI to read the logs of the `calico-node` and `kube-flannel` containers on the RHEL 8.4 nodes (the UI says "disconnected")
6) Observe that when using `docker logs k8s_kube-flannel_canal-XXXX` to read the container logs on the RHEL 8.4 nodes you see:
```
E0901 16:46:29.953388       1 iptables.go:115] Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t nat -C POSTROUTING ! -s 10.42.0.0/16 -d 10.42.8.0/24 -j RETURN --wait]: exit status -1:
E0901 17:02:11.148676       1 iptables.go:115] Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t nat -C POSTROUTING -s 10.42.0.0/16 ! -d 224.0.0.0/4 -j MASQUERADE --random-fully --wait]: exit status -1:
E0901 17:42:27.498460       1 iptables.go:115] Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t filter -C FORWARD -d 10.42.0.0/16 -j ACCEPT --wait]: exit status -1:
E0901 19:20:11.934311       1 iptables.go:115] Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t filter -C FORWARD -d 10.42.0.0/16 -j ACCEPT --wait]: exit status -1:
E0901 20:15:59.272599       1 iptables.go:115] Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t filter -C FORWARD -s 10.42.0.0/16 -j ACCEPT --wait]: exit status -1:
```
7) Observe in dmesg on the nodes that iptables commands are segfaulting:
```
[Wed Sep  1 17:10:11 2021] iptables-save[1024483]: segfault at 80 ip 00007f39cff4a964 sp 00007fffd5f6d8a8 error 4 in libnftnl.so.11.2.0[7f39cff45000+19000]
[Wed Sep  1 17:10:11 2021] Code: 83 c4 08 5b 5d 41 5c 41 5d c3 0f 1f 40 00 48 83 c4 08 31 c0 5b 5d 41 5c 41 5d c3 66 66 2e 0f 1f 84 00 00 00 00 00 f3 0f 1e fa <48> 8b 87 80 00 00 00 48 83 ef 80 48 39 f8 74 1b 85 f6 75 0c eb 18
```

**Result**
The overlay network is not able to correctly pass traffic because the underlying iptables rules are not properly put into place.

**Expected Result**
iptables commands run from within `rancher/coreos-flannel:v0.13.0-rancher1` should not segfault.

**Additional context**
This is discussed by the flannel project here:
https://github.com/flannel-io/flannel/issues/1408

A PR was merged into flannel on June 28:
https://github.com/flannel-io/flannel/pull/1449

SURE-3096

**Comment by jgallucci32** [comment]
I confirm the same issue here with the following setup
RHEL 8.3 4.18.0-240.22.1.el8_3.x86_64
Kubernetes version: 1.20.10

**Comment by StrongMonkey** [comment]
SURE-3096
SURE-3027
SURE-3028

**Comment by batiati** [comment]
Same issue here, in a newly deployed cluster

AlmaLinux 8.4 4.18.0-305.19.1.el8_4.x86_64
Kubernetes v1.21.5-rancher1-1

**Comment by StrongMonkey** [comment]
After my initial investigation and testing, flannel v0.15.1(v0.15.0) doesn't seem to fix this problem. The error is gone in flannel pod but segment error is still seen in kernel logs. Need to investigate more.

**Comment by StrongMonkey** [comment]
https://github.com/flannel-io/flannel/issues/1492 Might be releated that flannel v0.15.0 doesn't fix the issue.

**Comment by stefa975** [comment]
Hi,
I have done the same as you,
I found out the it might be the hyperkube image running kubelet and that is based on ubunu 20.04

**Comment by aiyengar2** [comment]
KDM branch to test on is `dev-v2.5-flannel-v0.15`.

**Comment by Auston-Ivison-Suse** [comment]
**Further testing**
- Rancher version: v2.5-head(354cf06)
- Docker install

Cluster info
- custom cluster
- 3 nodes: 1 worker, 1 etcd, 1 control plane
- ec2 instances: rhel-8.4-docker-20.10.7-selinux-on-rpm

**Prerequisites Before Testing**
1. Run the following command within all 3 of the nodes with the RHEL amis, _before_ attempting to provision the custom cluster.

`sudo systemctl disable nm-cloud-setup.service nm-cloud-setup.timer`

2. In your rancher setup: go to Global --> settings --> scroll to rke-metadata-config and insert:

` https://releases.rancher.com/kontainer-driver-metadata/dev-v2.5-flannel-v0.15`

**Reproduction steps** _Same as my previous posting_
1. Within the custom cluster deploy a workload in the default.
2. From explorer --> pods --> View the logs of pod with canal deployed on it in the namespace `kube-system`
3. You will see the following: 
` Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t filter -C FORWARD`
4. You will witness a segfault by running the command `dmesg` within the shell of the worker node.
For example ` segfault at 80 ip xxxxxx sp xxxxxxx  error 4 in libnftnl.so.11.2.0[7xxxxxxx]`
5. You can also check within the docker container by running `docker logs k8s_kube-flannel_canal-XXXX`
the output will contain the following : ` Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running [/sbin/iptables -t filter -C FORWARD -s xxxxx/xx -j ACCEPT --wait]: exit status -1: `

**Comment by aiyengar2** [comment]
@Auston-Ivison-Suse re: step 4, the result of the `dmesg` command on the shell of the worker node should not impact this issue; it means that the version of the `iptables` binary installed on the **host** is <1.8.5, which should not impact the ability for Flannel to set up overlay networking.

Also, regardless of the errors logged, is overlay networking still established? I tested this previously in CentOS 8 nodes in DO and did not experience any issues, so I'm wondering if this could be related to the specific AMI that you are testing with as I did not reproduce any errors on previous testing.

**Comment by aiyengar2** [comment]
I investigated @Auston-Ivison-Suse's setup to see where the identified segfaults are coming from.

**The TLDR is that the segfault that you are seeing is likely not related to Flannel or this issue.**

When I sshed into the canal-node pod and hopped into the kube-flannel container to run the following command, I saw the segfault in `libnftnl.so.11.2.0`:

```bash
bash-5.1# dmesg | grep "iptables"
[ 7185.469177] iptables[64524]: segfault at 80 ip 00007f00bd0ed964 sp 00007fff528b1b78 error 4 in libnftnl.so.11.2.0[7f00bd0e8000+19000]
```

However, on trying to identify the version of libnftnl installed on the Flannel container, I saw that the version installed was `libnftnl.so.11.4.0`, which seems to indicate the segfault is **not** coming from Flannel:

```bash
bash-5.1# ls /usr/lib | grep "libnftnl"
libnftnl.so.11
libnftnl.so.11.4.0
```

On the other hand, these segfaults do match with the version of libnftnl installed on the kube-proxy / kubelet containers based on rancher/hyperkube:v1.20.12-rancher1 as seen below, so it's likely that that is where the segfaults are coming from:

```bash
$ docker run -it --entrypoint /bin/bash rancher/hyperkube:v1.20.12-rancher1
root@d98ccd79cc35:/# ls /lib/x86_64-linux-gnu | grep "libnftnl"
libnftnl.so.11
libnftnl.so.11.2.0
```

This seems to be because the hyperkube images used by Rancher still have not been upgraded to using an iptables version >=1.8.5, for which I have opened an issue in https://github.com/rancher/rancher/issues/35709. Since those components use a similar mechanism to identify which packet filter to use to modify the iptables rules, it's possible that a similar error can be encountered on those containers that could prevent Kubernetes networking from being established, but the fix for that issue is independent of this since the current base image used (Ubuntu 20.04) does not seem to have a version available yet that has iptables >=1.8.5.

**Comment by Auston-Ivison-Suse** [comment]
**Validation Steps**
- Rancher version: v2.5-head(354cf06)
- Docker install

Cluster info
- custom cluster
- 3 nodes: 1 worker, 1 etcd, 1 control plane
- ec2 instances: rhel-8.4-docker-20.10.7-selinux-on-rpm
- k8s: v1.20.12-rancher1-2 (This is achieved, in part, by following the steps in the _Prerequisites_ section below)

**Prerequisites Before Testing**
1. Run the following command within all 3 of the nodes with the RHEL amis, _before_ attempting to provision the custom cluster.

`sudo systemctl disable nm-cloud-setup.service nm-cloud-setup.timer`

2. In your rancher setup: go to Global --> settings --> scroll to rke-metadata-config and insert:

` https://releases.rancher.com/kontainer-driver-metadata/dev-v2.5-flannel-v0.15`

3. Once the above meta data was changed we can then go to Global --> Tools --> Drivers where we will land on the Cluster Driver's page and refresh the Kubernetes Metadata. 

4. Now edit the config of the cluster to upgrade it to k8s: v1.20.12-rancher1-2. 

**Validation Steps** 
1. Within the custom cluster deploy a workload in the default namespace.
2. From explorer --> pods --> in the kube-system namespace --> Execute shell in one of the Canal pods.
3 . Change the container to kube-flannel, and run the `dmesg` command. 

**Result**
Within the output you will see something similar to:
`[ 7185.469177] iptables[64524]: segfault at 80 ip 00007f00bd0ed964 sp 00007fff528b1b78 error 4 in libnftnl.so.11.2.0[7f00bd0e8000+19000]`

For the purpose of this ticket, flannel is absolved of being the root cause. The above segfault seems to be tied to libnftnl.so.11.2.0/iptables version. 

@aiyengar2 has very kindly outlined more in his previous comment. We have created another issue to remove any segfault entirely: https://github.com/rancher/rancher/issues/35709.

**Comment by rajha-korithrien** [comment]
@aiyengar2 and @Auston-Ivison-Suse I appreciate the analysis and follow the logic as to why the segfault is not from flannel but instead from hyperkube. With that logic in mind, does it explain the fact that the logs in the flannel container say things like `Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running...`?

**Comment by Auston-Ivison-Suse** [comment]
@rajha-korithrien the error `Failed to ensure iptables rules: Error checking rule existence: failed to check rule existence: running...` was actually due to cluster being on the wrong k8s version, so once I upgraded to v1.20.12-rancher1-2 this disappeared.

**Comment by rimclellan** [comment]
Confirmed this still appears to be an issue:

RHEL 8.4
Kernel - 4.18.0-305.19.1.el8_4.x86_64 #1 SMP Tue Sep 7 07:07:31 EDT 2021 x86_64 x86_64 x86_64 GNU/Linux
Rancher v 2.6.2
Hyberkube- v1.20.13-rancher1 and v1.21.5-rancher1-1
Flannel - v0.15.1

**Comment by snasovich** [comment]
Assigning to myself for tracking purposes.

**Comment by jmcsagdc** [comment]
opened 2.6 version of this ticket here: https://github.com/rancher/rancher/issues/36115

**Comment by snasovich** [comment]
Ready to test based on https://github.com/rancher/rancher/issues/36021 follow-up issue addressed.

**Comment by Auston-Ivison-Suse** [comment]
**Validation Steps**
- Rancher version: v2.5-head(bd840b4)
- ha deploy

Cluster info
- custom cluster
- 8 nodes: 3 worker, 3 etcd, 2 control plane
- ec2 instances: rhel-8.4-docker-20.10.7-selinux-on-rpm
- k8s: v1.20.14-rancher2-1 

**Validation Steps** 
1. Within the custom cluster deploy a workload in the default namespace.

2. From explorer --> pods --> in the kube-system namespace --> Execute shell in one of the Canal pods.

3 . Change the container to kube-flannel, and run the `dmesg` command. 

**Result**
After going through the logs of the pod in the kube-system there should be no segfaults/issues with the ip-tables. This was an issue with the ubuntu image in hyper-kube, which has been updated, as mentioned in issue [#36021](https://github.com/rancher/rancher/issues/36021)

**Comment by Auston-Ivison-Suse** [comment]
@Sahota1225 There was some miscommunication between my manager and me. I am told we are able to move this to done and this no longer needs verification from customer support.

**Comment by sowmyav27** [comment]
Reopening for an upgrade check. 

@Auston-Ivison-Suse as discussed offline, please validate an update usecase

**Comment by sowmyav27** [comment]
Closng this and reopening this one instead https://github.com/rancher/rancher/issues/36021
